junit-platform-commons/src/main/java/org/junit/platform/commons/util/ExceptionUtils.java

```diff
@@ -80,4 +94,77 @@ public static String readStackTrace(Throwable throwable) {
 		return stringWriter.toString();
 	}
 
+	/**
+	 * Prune the stack trace of the supplied {@link Throwable} by filtering its
+	 * elements using the supplied {@link Predicate}, except for
+	 * {@code org.junit.jupiter.api.Assertions} and
+	 * {@code org.junit.jupiter.api.Assumptions} that will always remain
+	 * present.
+	 *
+	 * <p>Additionally, all elements prior to and including the first
+	 * JUnit Launcher call will be removed.
+	 *
+	 * @param throwable the {@code Throwable} whose stack trace should be
+	 * pruned; never {@code null}
+	 * @param stackTraceElementFilter the {@code Predicate} used to filter
+	 * elements of the stack trace; never {@code null}
+	 *
+	 * @since 5.10
+	 */
+	public static void pruneStackTrace(Throwable throwable, Predicate<String> stackTraceElementFilter) {
+		Preconditions.notNull(throwable, "Throwable must not be null");
+		Preconditions.notNull(stackTraceElementFilter, "Predicate must not be null");
+
+		List<StackTraceElement> stackTrace = Arrays.asList(throwable.getStackTrace());
+		List<StackTraceElement> prunedStackTrace = new ArrayList<>();
+
+		Collections.reverse(stackTrace);
+
+		for (StackTraceElement element : stackTrace) {
+			String className = element.getClassName();
+			if (className.startsWith(JUNIT_PLATFORM_LAUNCHER_PACKAGE_PREFIX)) {
+				prunedStackTrace.clear();
+			}
+			else if (stackTraceElementFilter.test(className)
+					|| ALWAYS_INCLUDED_STACK_TRACE_ELEMENTS.contains(className)) {
+				prunedStackTrace.add(element);
+			}
+		}
+
+		Collections.reverse(prunedStackTrace);
+		throwable.setStackTrace(prunedStackTrace.toArray(new StackTraceElement[0]));
+	}
+
+	/**
+	 * Find all causes and suppressed exceptions in the backtrace of the
+	 * supplied {@link Throwable}.
+	 *
+	 * @param rootThrowable the {@code Throwable} to explore; never {@code null}
+	 * @return an immutable list of all throwables found, including the supplied
+	 * one; never {@code null}
+	 *
+	 * @since 5.10
+	 */
+	public static List<Throwable> findNestedThrowables(Throwable rootThrowable) {
+		Preconditions.notNull(rootThrowable, "Throwable must not be null");
+
+		HashSet<Throwable> visited = new HashSet<>();
```

Comment:
Preserving the order is not necessary for our use case but might make is easier to debug and reason about the code.

```suggestion
		Set<Throwable> visited = new LinkedHashSet<>();
```